src/biogeophys/LakeFluxesMod.F90

```diff
@@ -324,20 +336,62 @@ subroutine LakeFluxes(bounds, num_lakec, filter_lakec, num_lakep, filter_lakep,
             z0qg(p) = max(z0qg(p), minz0lake)
             z0hg(p) = max(z0hg(p), minz0lake)
          else if (snl(c) == 0) then    ! frozen lake with ice
-            z0mg(p) = z0frzlake
-            z0hg(p) = z0mg(p) / exp(params_inst%a_coef * (ust_lake(c) * z0mg(p) / 1.5e-5_r8)**params_inst%a_exp) ! Consistent with BareGroundFluxes
+            select case (z0param_method)
+            case ('Meier2022') 
+               z0mg(p) = params_inst%zglc
+
+
+               z0hg(p) = 70._r8 * nu_param / ust_lake(c) ! For initial guess assume tstar = 0
+
+            case ('ZengWang2007')
+               z0mg(p) = z0frzlake
+               z0hg(p) = z0mg(p) / exp(params_inst%a_coef * (ust_lake(c) * z0mg(p) / nu_param)**params_inst%a_exp) ! Consistent with BareGroundFluxes
+            end select
             z0qg(p) = z0hg(p)
          else                          ! use roughness over snow as in Biogeophysics1
-            z0mg(p) = params_inst%zsno
-            z0hg(p) = z0mg(p) / exp(params_inst%a_coef * (ust_lake(c) * z0mg(p) / 1.5e-5_r8)**params_inst%a_exp)  ! Consistent with BareGroundFluxes
+            select case (z0param_method)
+            case ('Meier2022') 
+               if(use_z0m_snowmelt) then
+                  if ( snomelt_accum(c) < 1.e-5_r8 )then
+                     z0mg(p) = exp(-b1_param * rpi * 0.5_r8 + b4_param) * 1.e-3_r8
+                  else
+                     z0mg(p) = exp(b1_param * (atan((log10(snomelt_accum(c))+0.23_r8)/0.08_r8)) + b4_param) * 1.e-3_r8
+                  end if
+                  
+               else
+                  z0mg(p) = params_inst%zsno
+                  
+               end if                       
+               z0hg(p) = 70._r8 * nu_param / ust_lake(c) ! For initial guess assume tstar = 0
+
+            case ('ZengWang2007')
+               if(use_z0m_snowmelt) then
+                  if ( snomelt_accum(c) < 1.e-5_r8 ) then
+                     z0mg(p) = exp(-b1_param * rpi * 0.5_r8 + b4_param) * 1.e-3_r8
+                  else
+                     z0mg(p) = exp(b1_param * (atan((log10(snomelt_accum(c))+0.23_r8)/0.08_r8)) + b4_param) * 1.e-3_r8
+                  end if
+               else
+                  z0mg(p) = params_inst%zsno
+               end if                      
+               z0hg(p) = z0mg(p) / exp(params_inst%a_coef * (ust_lake(c) * z0mg(p) / nu_param)**params_inst%a_exp) ! Consistent with BareGroundFluxes
```

Comment:
I think I can also reduce repetition in the above with bfb results.

Reply:
Testing the refactors here and in FrictionVelocityMod with the test that I used in [this earlier post.](https://github.com/ESCOMP/CTSM/pull/2045#issuecomment-1636423638)